src/login/challenge/email-verification.ts

```diff
@@ -0,0 +1,149 @@
+import { AbstractLoginChallenge } from './abstract.js';
+import { AuthorizationChallengeRequest } from '../types.js';
+import { A12nLoginChallengeError } from '../error.js';
+import * as services from '../../services.js';
+import { PrincipalIdentity } from '../../types.js';
+import { BadRequest } from '@curveball/http-errors';
+import { markVerified } from '../../principal-identity/service.js';
+
+type EmailVerification = {
+  email_verification_code: string;
+}
+
+
+/**
+ * Password-based authentication strategy.
+ */
+export class LoginChallengeEmailVerification extends AbstractLoginChallenge<EmailVerification> {
+
+  /**
+   * The type of authentication factor this class provides.
+   */
+  readonly authFactor = 'email-verification';
+
+  /**
+  * Checks if the user has an unverified email.
+  * Returns true if an unverified email identity is found,
+  */
+
+  async userHasChallenge(): Promise<boolean> {
+
+    const identity = await this.findUnverifiedEmailIdentity();
+    return identity !== null;
```

Comment:
So one way this Challenge has to behave slightly different from the other one, is that this should check the _specific_ identity (username) that the user used to login, whereas the email-otp will accept any.

The problem you have here is that you don't have access to this in this function. I think the AbstractLoginChallenge class should be modified to add 1 more constructor argument, so the signature becomes:

```
constructor(principal: User, identity: PrincipalIdentity, logger: UserEventLogger, ip: string) {
```

Then `userHasChallenge` can operate on `this.identity`.

This will then make a few helper functions in this class also unneeded.

The logic is that any account can have 1 or more identities associated to it, but only a verified identity can be used to log in.